lib/src/Uncertainty/Algorithm/WeightedExperiments/TensorProductExperiment.cxx

```diff
@@ -0,0 +1,192 @@
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "openturns/TensorProductExperiment.hxx"
+#include "openturns/Tuples.hxx"
+#include "openturns/Exception.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+#include "openturns/Log.hxx"
+#include "openturns/BlockIndependentDistribution.hxx"
+#include <iostream>
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(TensorProductExperiment)
+
+typedef Collection< Point > PointCollection;
+
+/* Default constructor */
+TensorProductExperiment::TensorProductExperiment()
+  : WeightedExperimentImplementation()
+  , collection_(0)
+  , nodes_(0, 0)
+  , isAlreadyComputedNodesAndWeights_(false)
+{
+  // Nothing to do
+}
+
+/* Constructor with parameters */
+TensorProductExperiment::TensorProductExperiment(const WeightedExperimentCollection & collection)
+  : WeightedExperimentImplementation()
+  , collection_(collection)
+  , nodes_(0, 0)
+  , isAlreadyComputedNodesAndWeights_(false)
+{
+  const UnsignedInteger numberOfMarginalExperiments = collection_.getSize();
+  BlockIndependentDistribution::DistributionCollection distributionCollection(numberOfMarginalExperiments);
+  UnsignedInteger size = 1;
+  for (UnsignedInteger i = 0; i < numberOfMarginalExperiments; ++i)
+  {
+    size *= collection_[i].getSize();
+    distributionCollection[i] = collection_[i].getDistribution();
+  }
+  WeightedExperimentImplementation::setSize(size);
+  const BlockIndependentDistribution distribution(distributionCollection);
+  WeightedExperimentImplementation::setDistribution(distribution);
+}
+
+/* Virtual constructor */
+TensorProductExperiment * TensorProductExperiment::clone() const
+{
+  return new TensorProductExperiment(*this);
+}
+
+/* String converter */
+String TensorProductExperiment::__repr__() const
+{
+  OSS oss;
+  oss << "class=" << GetClassName()
+    << " name=" << getName()
+    << " collection=" << collection_;
+  return oss;
+}
+
+Bool TensorProductExperiment::hasUniformWeights() const
+{
+  const UnsignedInteger numberOfMarginalExperiments = collection_.getSize();
+  bool hasUniformWeights = true;
+  for (UnsignedInteger i = 0; i < numberOfMarginalExperiments; ++i)
+  {
+    if (!collection_[i].hasUniformWeights())
+    {
+      hasUniformWeights = false;
+      break;
+    }
+  }
+
+  return hasUniformWeights;
+}
+
+/* Sample generation */
+Sample TensorProductExperiment::generateWithWeights(Point & weights) const
+{
+  LOGDEBUG(OSS() << "generateWithWeights()");
+  LOGDEBUG(OSS() << "isAlreadyComputedNodesAndWeights_ = " << isAlreadyComputedNodesAndWeights_);  
+  if (!isAlreadyComputedNodesAndWeights_) computeNodesAndWeights();
```

Comment:
I think we should avoid this otherwise the `generate` method will produce the same experiment at each call.
Isn't it?

Reply:
This isn't quite right.
- If the `setWeightedExperimentCollection` method is used, then the `isAlreadyComputedNodesAndWeights_` attribute is set to false, and the sample is updated. 
- Otherwise, then the `generate()` method will, indeed, produce the same sample at each call. This is the expected result, given that there is nothing random in this experiment. Even if we pass a `MonteCarloExperiment` to it, the tensorization algorithm will produce the same sample, which is the intended result, at least in my opinion. 

Does this fit in the general framework for Experiments?

Reply:
It is not the expected behavior, as you can check it using any of the random weighted experiments. Otherwise, you are just implementing a composition method for weighted samples, and a weighted experiment is more than that: it is also a generator. Suppose that MonteCarloExperiment returns the exact same sample each time one call its generate() method, then the ProbabilitySimulationAlgorithm becomes useless!

Reply:
I understand. Here is the script I used to check this;
```python
import openturns as ot

experiment1 = ot.MonteCarloExperiment(ot.Normal(1), 2)
experiment2 = ot.MonteCarloExperiment(ot.Normal(1), 2)
collection = [experiment1, experiment2]
multivariateExperiment = ot.TensorProductExperiment(collection)
nodes, weights = multivariateExperiment.generateWithWeights()
print("Call #1")
print(nodes)
print("Call #2")
nodes, weights = multivariateExperiment.generateWithWeights()
print(nodes)
```
and here is the result after the fix.
```
$ python3 ~/Documents/test_MonteCarloTensorProduct.py 
Call #1
0 : [  0.608202 -0.438266 ]
1 : [ -1.26617  -0.438266 ]
2 : [  0.608202  1.20548  ]
3 : [ -1.26617   1.20548  ]
Call #2
0 : [ -2.18139  -0.355007 ]
1 : [  0.350042 -0.355007 ]
2 : [ -2.18139   1.43725  ]
3 : [  0.350042  1.43725  ]
```
It is fixed. Thank for for watching that!